src/bindings/swig/python/tests/test_gopts.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/python3
+
+import sys
+import kdb
+import unittest
+from ctypes import c_char_p
+
+base_key = "/sw/org/kdbdummy/#0/current"
+spec_base_key = "spec:" + base_key
+
+spec = kdb.KeySet(10, kdb.Key(spec_base_key, kdb.KEY_META, "command", ""),
+		  kdb.Key(spec_base_key + "/printversion",
+			  kdb.KEY_META, "description", "print version information and exit (ignoring all other options/commands/parameters)",
+			  kdb.KEY_META, "opt", "v",
+			  kdb.KEY_META, "opt/arg", "none",
+			  kdb.KEY_META, "opt/long", "version"),
+		  kdb.Key(spec_base_key + "/getter",
+			  kdb.KEY_META, "description", "get a key's value",
+			  kdb.KEY_META, "command", "get"),
+		  kdb.Key(spec_base_key + "/getter/verbose",
+			  kdb.KEY_META, "description", "print additional information about where the value comes from",
+			  kdb.KEY_META, "opt", "v",
+			  kdb.KEY_META, "opt/long", "verbose",
+			  kdb.KEY_META, "opt/arg", "none"),
+		  kdb.Key(spec_base_key + "/getter/keyname",
+			  kdb.KEY_META, "description", "name of the key to read",
+			  kdb.KEY_META, "args", "indexed",
+			  kdb.KEY_META, "args/index", "0"),
+		  kdb.Key(spec_base_key + "/setter",
+			  kdb.KEY_META, "description", "set a key's value",
+			  kdb.KEY_META, "command", "set"),
+		  kdb.Key(spec_base_key + "/setter/verbose",
+			  kdb.KEY_META, "description", "print additional information about where the value will be stored",
+			  kdb.KEY_META, "opt", "v",
+			  kdb.KEY_META, "opt/long", "verbose",
+			  kdb.KEY_META, "opt/arg", "none"),
+		  kdb.Key(spec_base_key + "/setter/keyname",
+			  kdb.KEY_META, "description", "name of the key to write",
+			  kdb.KEY_META, "args", "indexed",
+			  kdb.KEY_META, "args/index", "0"),
+		  kdb.Key(spec_base_key + "/setter/value",
+			  kdb.KEY_META, "description", "value to be written",
+			  kdb.KEY_META, "args", "indexed", kdb.KEY_META, "args/index", "1"),
+		  kdb.Key(spec_base_key + "/dynamic/#",
+			  kdb.KEY_META, "description", "dynamically call a user-supplied command",
+			  kdb.KEY_META, "args", "remaining")
+		  )
+
+def setupSpec():
+	with kdb.KDB() as db:
+		ks = kdb.KeySet(10)
+		db.get(ks, spec_base_key)
+		if len(ks.cut(kdb.Key(spec_base_key))) > 0:
+			print("ERROR: Couldn't setup spec, keys exist!", file=sys.stderr)
+			exit(1)
+
+		ks.extend(spec)
+		db.set(ks, spec_base_key)
+
+def removeSpec():
+	with kdb.KDB() as db:
+		ks = kdb.KeySet(10)
+		db.get(ks, spec_base_key)
+		ks.cut(kdb.Key(spec_base_key))
+		db.set(ks, spec_base_key)
+
+def array_index(index):
+	return "_" * (len(str(index)) - 1) + str(index)
+
+class GOpts(unittest.TestCase):
+	def test_gopts(self):
+		try:
+			setupSpec()
+
+			custom_argv = ["test", "get", "-v", "user:/"]
+			custom_envp = []
+
+			# FIXME: somehow convert python string list into const char * const * with NULL at the end
+			c_argv = custom_argv
+			c_envp = custom_envp
```

Comment:
I have no idea how to make this work... I did a quick google search, and
- it is quite easy to call a function with a `std::vector<std::string>` argument
- we can put some code into `kdb.i` to make SWIG convert the string list into `char **` (see #3608)

But both of these "solutions", don't work, because we need a `char **` that lives until after `kdbClose()` is called. Everything I found just creates a temporary allocation that is freed right after the function call that needs the `char **` argument.

You can also use `ctypes` to create a `char **` and fill it from Python, but I couldn't figure out how to pass that to `kdb::goptsContract` via SWIG.

I didn't even look into the other SWIG bindings...

PS. Not quite sure how Python's GC works, but if it tries to garbage collect the arrays to early, we can probably just put something like `custom_argv.append("")` right at the end, so it doesn't think the array isn't needed anymore.

Reply:
Found a solution, for using `std::vector<std::string>`